tests/frontend/test_engine.py

```diff
@@ -166,209 +174,304 @@ def inspect():
         assert inspect() == expected2
 
 
-class TestStarshipEngine:
-    """
-    Tests various methods on the remote engine StarshipEngine.
+@pytest.fixture
+def connection():
+    """A mock connection object."""
+    return Connection(token="token", host="host", port=123, use_ssl=True)
+
+
+def mock_return(return_value):
+    """A helper function for defining a mock function that returns the given value for
+    any arguments.
     """
+    return lambda *args, **kwargs: return_value
+
+
+class MockResponse:
+    """A mock response with a JSON or binary body."""
+
+    def __init__(self, status_code, json_body=None, binary_body=None):
+        self.status_code = status_code
+        self.json_body = json_body
+        self.binary_body = binary_body
+
+    def json(self):
+        return self.json_body
+
+    @property
+    def content(self):
+        return self.binary_body
 
-    def test_init(self, monkeypatch):
-        """
-        Tests that a StarshipEngine instance is correctly initialized when additional APIClient
-        parameters are passed.
-        """
-        mock_api_client = MagicMock()
-        monkeypatch.setattr("strawberryfields.engine.APIClient", mock_api_client)
-        engine = StarshipEngine("chip0")
-        assert engine.client == mock_api_client()
-        assert engine.jobs == []
-        assert engine.REMOTE == True
-
-    def test_reset(self, starship_engine):
-        """
-        Tests that StarshipEngine.jobs is correctly cleared when callling StarshipEngine.reset.
-        """
-        starship_engine.jobs.append(MagicMock())
-        assert len(starship_engine.jobs) == 1
-        starship_engine.reset()
-        assert len(starship_engine.jobs) == 0
 
-    def test__get_blackbird(self, starship_engine, monkeypatch):
+class MockServer:
+    """A mock platform server that fakes a processing delay by counting requests."""
+
+    REQUESTS_BEFORE_COMPLETED = 3
+
+    def __init__(self):
+        self.request_count = 0
+
+    def get_job_status(self, _id):
+        """Returns a 'queued' job status until the number of requests exceeds a defined
+        threshold, beyond which a 'complete' job status is returned.
         """
-        Tests that StarshipEngine._get_blackbird returns the correct string given name,
-        shots, and program parameters.
+        self.request_count += 1
+        return (
+            JobStatus.COMPLETED
+            if self.request_count >= self.REQUESTS_BEFORE_COMPLETED
+            else JobStatus.QUEUED
+        )
+
+
+class TestResult:
+    """Tests for the ``Result`` class."""
+
+    def stateless_result_raises_on_state_access(self):
+        """Tests that `result.state` raises an error for a stateless result.
         """
-        methods = MagicMock()
-        inputs = MagicMock()
+        result = Result([[1, 2], [3, 4]], is_stateful=False)
+
+        with pytest.raises(AttributeError):
```

Comment:
It would be good to also include the `match=` argument here, just to make sure that the _right_ attribute error is raised!

Reply:
Thanks, I've added `match` checks to all the relevant tests.